cardano-explorer-node/src/Explorer/Node/Database.hs

```diff
@@ -0,0 +1,111 @@
+import qualified Control.Concurrent.STM.TBQueue as TBQ
+
+import qualified Explorer.DB as DB
+import           Explorer.Node.Insert
+import           Explorer.Node.Metrics
+import           Explorer.Node.Rollback
+import           Explorer.Node.Util
+
+import           Ouroboros.Consensus.Ledger.Byron (ByronBlockOrEBB (..))
+import           Ouroboros.Consensus.Ledger.Byron.Config (ByronConfig)
+import           Ouroboros.Network.Block (BlockNo (..), Point (..))
+
+import qualified System.Metrics.Prometheus.Metric.Gauge as Gauge
+
+
+data NextState
+  = Continue
+  | Done
+
+data DbAction
+  = DbApplyBlock !(ByronBlockOrEBB ByronConfig) !BlockNo
+  | DbRollBackToPoint !(Point (ByronBlockOrEBB ByronConfig))
+  | DbFinish
+
+newtype DbActionQueue = DbActionQueue
+  { dbActQueue :: TBQueue DbAction
+  }
+
+lengthDbActionQueue :: DbActionQueue -> STM Natural
+lengthDbActionQueue (DbActionQueue q) = STM.lengthTBQueue q
+
+newDbActionQueue :: IO DbActionQueue
+newDbActionQueue = DbActionQueue <$> TBQ.newTBQueueIO 2000
+
+writeDbActionQueue :: DbActionQueue -> DbAction -> STM ()
+writeDbActionQueue (DbActionQueue q) = TBQ.writeTBQueue q
+
+
+runDbThread :: Trace IO Text -> Metrics -> DbActionQueue -> IO ()
+runDbThread trce metrics queue = do
+    logInfo trce "Running DB thread"
+    loop
+    mBlkNo <-  DB.runDbNoLogging DB.queryLatestBlockNo
+    case mBlkNo of
+      Nothing -> pure ()
+      Just blkNo -> Gauge.set (fromIntegral blkNo) $ mDbHeight metrics
+  where
+    loop = do
+      xs <- blockingFlushDbActionQueue queue
+      logDebug trce $ "runDbThread: " <> textShow (length xs) <> " blocks"
+      nextState <- runActions trce xs
+      case nextState of
+        Continue -> loop
+        Done -> pure ()
+
+-- | Run the list of 'DbAction's. Block are applied in a single set (as a transaction)
+-- and other operations are applied one-by-one.
+runActions :: Trace IO Text -> [DbAction] -> IO NextState
+runActions trce =
+    dbAction Continue
+  where
+    dbAction :: NextState -> [DbAction] -> IO NextState
+    dbAction next [] = pure next
+    dbAction Done _ = pure Done
+    dbAction Continue xs =
+      case spanDbApply xs of
+        ([], DbFinish:_) -> do
+            pure Done
+        ([], DbRollBackToPoint pt:ys) -> do
+            rollbackToPoint trce pt
+            -- After rollback, a roll forward will happen so quickly its not worth
+            -- updating the metric.
+            dbAction Continue ys
+        (ys, zs) -> do
+          insertByronBlockOrEbbList trce ys
+          if null zs
+            then pure Continue
+            else dbAction Continue zs
+
+-- | Block if the queue is empty and if its not read/flush everything.
+-- Use this instead of STM.check to make sure it blocks if the queue is empty.
+blockingFlushDbActionQueue :: DbActionQueue -> IO [DbAction]
+blockingFlushDbActionQueue (DbActionQueue queue) = do
+  STM.atomically $ do
+    x <- TBQ.readTBQueue queue
+    xs <- TBQ.flushTBQueue queue
```

Comment:
why read it in 2 pieces and join them together?

Reply:
That is explained in the comment above the function. The flush operation does not block and may return an empty list so using when it just returns an empty list will just busy spin. This version block until the queue is non-empty.

Reply:
ahhh, i didnt notice that detail when i was writing the code
so that results in my old version opening up a whole psql transaction to do nothing!